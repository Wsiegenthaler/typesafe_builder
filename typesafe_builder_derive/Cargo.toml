[package]
name = "typesafe_builder_derive"
version = "0.1.0"
edition = "2024"
authors = ["tomoikey"]
readme = "README.md"
repository = "https://github.com/tomoikey/typesafe_builder"
categories = ["accessibility", "development-tools", "rust-patterns"]
license = "MIT"
description = "A procedural macro to generate type-safe builder patterns for Rust structs"

[lib]
proc-macro = true

[dependencies]
syn = { version = "2", features = ["full"] }
quote = "1"
darling = "0.20"

[dev-dependencies]
trybuild = "1.0.105"
