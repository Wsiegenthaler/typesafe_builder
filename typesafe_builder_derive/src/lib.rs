mod derive_builder;
mod input;

use darling::FromDeriveInput;
use derive_builder::derive_builder_impl;
use input::Input;
use proc_macro::TokenStream;
use syn::{DeriveInput, parse_macro_input};

#[proc_macro_derive(Builder, attributes(builder))]
pub fn derive_builder(tokens: TokenStream) -> TokenStream {
    let input = parse_macro_input!(tokens as DeriveInput);
    let input = match Input::from_derive_input(&input) {
        Ok(v) => v,
        Err(e) => return e.write_errors().into(),
    };

    match derive_builder_impl(input) {
        Ok(expanded) => TokenStream::from(expanded),
        Err(e) => TokenStream::from(e.write_errors()),
    }
}
