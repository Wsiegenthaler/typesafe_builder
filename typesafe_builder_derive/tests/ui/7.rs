use typesafe_builder_derive::Builder;

fn main() {
    #[derive(Builder)]
    struct User {
        #[builder(optional)]
        name: Option<String>,
        #[builder(required_if = "nonexistent_field")]
        age: Option<u8>,
    }
} 