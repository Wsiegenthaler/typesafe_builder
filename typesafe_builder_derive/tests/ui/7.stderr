error: Field 'nonexistent_field' used in condition does not exist
 --> tests/ui/7.rs:4:14
  |
4 |     #[derive(Builder)]
  |              ^^^^^^^
  |
  = note: this error originates in the derive macro `Builder` (in Nightly builds, run with -Z macro-backtrace for more info)
